### PR TITLE
Remove passwords from logged exception stack traces

### LINKED ISSUE #16318: Exception logger should sanitize potential sensitive data such as passwords
The exception logger should sanitize potential sensitive data such as the parameters passed to `OC_Login` and other functions.

In my opinion the best approach would be to go with a blacklist of functions and sanitize all parameters given to it. – We need however also apps to specify their own blacklisted function names as things might fail there as well :see_no_evil:

Thing is we simply don't know the user password and storing this in the PHP session is also a no-go from my opinion.

Excerpt from an affected log in the JSON error log.

```
OC\\\\User\\\\Session->login('test', 't3st')\\n#13 \\\/var\\\/www\\\/owncloud\\\/lib\\\/base.php(978): OC_User::login('test', 't3st')\\n#14 \\\/var\\\/www\\\/owncloud\\\/lib\\\/base.php(878): OC::tryFormLogin()\\n#15 \\\/var\\\/www\\\/owncloud\\\/lib\\\/base.php(840): OC::handleLogin()\\n#16 \\\/var\\\/www\\\/owncloud\\\/index.php(36): OC::handleRequest()\\n#17 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Driver\\\/AbstractMySQLDriver.php\",\"Line\":115
```

Tentatively assigning 8.1, if the required change set is too big I will move this over to 8.2

Alternative proposals? @DeepDiver1975, @MorrisJobke, @PVince81, others?

### PR DESCRIPTION
**Pull request by MorrisJobke**
- fixed #16318
- create logException in ILogger
- add unit tests

cc @DeepDiver1975 @nickvergessen @LukasReschke

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
:hust: unit test :hust:

File: lib/public/util.php [hunk 1]
```diff
@@ -168,6 +168,7 @@ public static function logException( $app, \Exception $ex, $level = \OCP\Util::F
 			'File' => $ex->getFile(),
 			'Line' => $ex->getLine(),
 		);
+		$exception['Trace'] = preg_replace('!(login|checkPassword)\(.*\)!', '$1(*** username and password replaced ***)', $exception['Trace']);
```
Comment (LukasReschke) [inline lib/public/util.php]: :spooky:

Comment (LukasReschke) [inline lib/public/util.php]: :beers:

**Comment by MorrisJobke** [comment]
> :hust: unit test :hust:

Choose one of the following:

* Nearly impossible - don't do it.
* Or I can throw an exception and then parse the logs.
* Or I create yet another private static method in OC_Util and test against this.

**Comment by LukasReschke** [comment]
@MorrisJobke I added :beers: to the radio box.

**Comment by DeepDiver1975** [comment]
I did add :pizza:

**Comment by MorrisJobke** [comment]
@DeepDiver1975 @LukasReschke I could also create a `logException` to the `ILogger` interface and unit test this

**Comment by LukasReschke** [comment]
>  I could also create a logException to the ILogger interface and unit test this

Sounds good for me.

**Comment by DeepDiver1975** [comment]
> Sounds good for me.

:+1:

**Comment by MorrisJobke** [comment]
I added unit tests and refactored everything to `logException` in `ILogger`. \o/

**Comment by MorrisJobke** [comment]
:pizza: and :beers: - I'm off for today :wave:

**Comment by LukasReschke** [comment]
Failing test is unrelated.

File: lib/private/log.php [hunk 2]
```diff
@@ -241,4 +241,25 @@ public function log($level, $message, array $context = array()) {
 			call_user_func(array($logger, 'write'), $app, $message, $level);
 		}
 	}
+
+	/**
+	 * Logs an exception very detailed
+	 *
+	 * @param \Exception $exception
+	 * @param array $context
+	 * @return null
```
Comment (LukasReschke) [inline lib/private/log.php]: Please use void or keep this empty. Null is something else. (http://phpdoc.org/docs/latest/references/phpdoc/tags/return.html)

File: lib/public/ilogger.php [hunk 3]
```diff
@@ -122,4 +122,14 @@ public function debug($message, array $context = array());
 	 * @since 7.0.0
 	 */
 	public function log($level, $message, array $context = array());
+
+	/**
+	 * Logs an exception very detailed
+	 *
+	 * @param \Exception $exception
+	 * @param array $context
+	 * @return null
```
Comment (LukasReschke) [inline lib/public/ilogger.php]: Please use void or keep this empty. Null is something else. (http://phpdoc.org/docs/latest/references/phpdoc/tags/return.html)

**Comment by MorrisJobke** [comment]
@LukasReschke Done ;)

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
@LukasReschke review please - thx

**Comment by LukasReschke** [comment]
:+1:

@DeepDiver1975 Merge please.

**Comment by LukasReschke** [comment]
@MorrisJobke :pizza: and :beer: please.

**Pull request merged**